keps/sig-node/0035-20190130-topology-manager.md

```diff
@@ -413,6 +446,34 @@ _Figure: Topology Manager hint provider registration._
 
 _Figure: Topology Manager fetches affinity from hint providers._
 
+### Deprecates the Policy Flag of Topology Manager
+Since the Pod Spec is able to specify preferred topology policy of a Pod
+and the Topology Manager is able to read `v1core.PodSpec.topologyPolicy` field when it is configured with the dynamic policy,
+the policy flag will be deprecated on Beta stage(v1.20) and the dynamic policy will be a default behavior of the Topology Manager.
+
+Before the deprecation of the policy flag,
+the dynamic policy should be configured on a node where operator expects the Topology Manager to get a topology policy of a pod from a Pod Spec.
+
+If topology policy is configured on the node as none/best-effort/restriced/single-numa-node, it takes precedence over the topology policy specified in Pod Spec.
+The behavior of Topology policies is listed below.
+- **none(default)**
+  * Kubelet does not consult with Topology Manager for placement decisions.
+- **best-effort/restricted/single-numa-node**
+  * If a node is configured with these policies, the Topology Manager will coordinate assignment of resources based on configured topology policy for all pods that were scheduled to the node.
+  * The Topology Manager will not read the `v1core.PodSpec.topologyPolicy` field of Pod Spec.
+- **dynamic**
+  * The Topology Manager reads `v1core.PodSpec.topologyPolicy` field to coordinate assignment of resources for a Pod.
+
+## Practical challenges
+Non-NUMA aware scheduler may bind a pod to a node, where topology requirements cannot be satisfied. Topology Manager will reject a pod admission with TopologyError and pod will result in a `Terminated` state on a node.
+
+#### Mitigations
+1. Using a controller that guarantees to bind a Pod to a node where the desired topology of Pod's resource assignment can be satisfied.
+    1. A new object is required to describe the desired topology, the object can be a new workload type `TopologySet` or CRD.
+2. A Topology-Aware Scheduler that would consider the topology of available resources on nodes.
+    1. Custom Scheduler could be provisioned to take care of the scheduling for high-performance applications. Detection of node's resources topology can be based on Node Feature Discovery with the usage of user-specified features.
```

Comment:
Regarding using Node Feature Discovery here, a couple of questions: my understanding is a custome scheduler should take into account not only the node resource topology (e.g. "this node has two NUMA nodes") but also the node resource availability (e.g. "this node has two NUMA nodes, and 4 cores free on NUMA node 0 and 6 core free on NUMA node 1") at the moment of the scheduling decision, right?

Reply:
Right, taking node resource availability into account would be very important during scheduling, this could prevent assigning a pod to the node where each particular NUMA node doesn't have enough CPUs  but in common they have enough.
Example here:
- pod request: 4 exclusive CPUs from one NUMA node
- NUMA 1: 3 free CPUs, NUMA 2: 2 free CPUs, Together: 5 free CPUs

Additionally `topologyPolicy` field specified in Pod Spec could be used to inform the scheduler if it should look on NUMA resources availability or not, depending on selected policy.

Reply:
Yep, hence my second question: is really NFD a good place to expose this information? Reasons I'm asking:
1. AFAIU node-feature-labels aren't supposed to change very often; surely the node allocation changes very frequently, so I wonder if this is the right approach to convey this information. Maybe we should move in the direction of the [prometheus node_exporter](https://github.com/prometheus/node_exporter) ?
2. did you planned already about how to extract the current allocation values (e.g. how many cores per-NUMA are actually being allocated) and expose them?

Last: while important in the big picture, I'm not sure this KEP PR is the right place to discuss this sub-topic, should we move in another venue?

Reply:
Here is an Issue #90204 created to discuss this:
https://github.com/kubernetes/kubernetes/issues/90204